src/Loader/ChainLoader.php

```diff
@@ -21,21 +21,30 @@
  */
 final class ChainLoader implements LoaderInterface
 {
+    /**
+     * @var \Traversable<LoaderInterface>|LoaderInterface[]
+     */
+    private $loaders;
+
+    /**
+     * @var array<string, bool>
+     */
     private $hasSourceCache = [];
-    private $loaders = [];
 
     /**
-     * @param LoaderInterface[] $loaders
+     * @param iterable<LoaderInterface> $loaders
      */
-    public function __construct(array $loaders = [])
+    public function __construct(iterable $loaders = [])
     {
-        foreach ($loaders as $loader) {
-            $this->addLoader($loader);
-        }
+        $this->loaders = $loaders;
     }
 
     public function addLoader(LoaderInterface $loader): void
     {
+        if (!\is_array($this->loaders)) {
+            $this->loaders = \iterator_to_array($this->loaders, false);
+        }
+
```

Comment:
Iterable (generators) can not be mutated by adding a new elements, you already know that. Order of loaders must be preserved. So adding loader with `addLoader()` breaks lazy loading - assuming that loaders are provided  as generator within constructor.

So this is simple implementation, but it has its own flaws.

There is, of course, better method to this madness :) -> to add some class, per example `LoadersCollection` where it would be allowed to add loader and/or  loaders without triggering generator and by preserving order.

Do you want that kind of implementation, perhaps? Or you are satisfied with this one?

Reply:
`addLoaders()` now supports preserving lazy loading of loaders provided as \Generator.

method `getLoaders()` will unpack loaders.